specifyweb/frontend/js_src/lib/components/WorkBench/WbView.tsx

```diff
@@ -252,867 +202,83 @@ export function WbViewReact({
-                ? dataset.rows.length
-                : hotRef.current!.hotInstance!.countRows() -
-                  hotRef.current!.hotInstance!.countEmptyRows(true)
-            }
+            hasUnsavedChanges={hasUnsavedChanges}
+            isUploaded={isUploaded}
+            onDatasetRefresh={triggerDatasetRefresh}
+            mappings={mappings!}
+            checkDeletedFail={checkDeletedFail}
+            onSpreadsheetUpToDate={spreadsheetUpToDate}
+            workbench={workbench}
+            onToggleResults={toggleResults}
           />
         </div>
-        <Button.Small
-          aria-haspopup="grid"
-          aria-pressed={showToolkit}
-          onClick={toggleToolkit}
-        >
-          {commonText.tools()}
-        </Button.Small>
-        <span className="-ml-1 flex-1" />
-        {canUpdate || isMapped ? (
-          <Link.Small href={`/specify/workbench/plan/${dataset.id}/`}>
-            {wbPlanText.dataMapper()}
-          </Link.Small>
-        ) : undefined}
-        <WbActionsComponent
-          dataset={dataset}
-          hasUnsavedChanges={hasUnsavedChanges}
-          isUploaded={isUploaded}
-          onRefresh={triggerRefresh}
-          mappings={mappings!}
-          checkDeletedFail={checkDeletedFail}
-          onSpreadsheetUpToDate={spreadsheetUpToDate}
-          workbench={workbench}
-          onToggleResults={toggleResults}
-        />
-      </div>
-      {showToolkit ? (
-        <WbToolkit
-          dataset={dataset}
-          hotRef={hotRef}
-          mappings={mappings!}
-          data={data}
-          handleDatasetDelete={handleDatasetDelete}
-          hasUnsavedChanges={hasUnsavedChanges}
-          triggerRefresh={triggerRefresh}
-        />
-      ) : undefined}
-      <div className="flex flex-1 gap-4 overflow-hidden">
-        <section className="flex-1 overflow-hidden overscroll-none">
-          <WbSpreadsheet
+        {showToolkit ? (
+          <WbToolkit
             dataset={dataset}
-            hotRef={hotRef}
-            isUploaded={isUploaded}
-            data={data}
-            validation={workbench.validation}
-            cells={workbench.cells}
-            disambiguation={workbench.disambiguation}
-            hooks={hooks}
+            hot={hot}
             mappings={mappings!}
+            data={data}
+            onDatasetDeleted={handleDatasetDeleted}
+            hasUnsavedChanges={hasUnsavedChanges}
+            triggerDatasetRefresh={triggerDatasetRefresh}
           />
-        </section>
-        {showResults ? (
-          <aside aria-live="polite">
-            <WbUploaded
-              dataSetId={dataset.id}
-              dataSetName={dataset.name}
+        ) : undefined}
+        <div className="flex flex-1 gap-4 overflow-hidden">
+          <section className="flex-1 overflow-hidden overscroll-none">
+            <WbSpreadsheet
+              dataset={dataset}
+              setHotTable={setHotTable as React.Ref<HotTable>}
```

Comment:
Getting this: `Type 'Dispatch<SetStateAction<HotTable | undefined>>' is not assignable to type '(instance: HotTable | null) => void'`

Reply:
looks like you need to change this line:

```diff
-  const [hotTable, setHotTable] = React.useState<HotTable>();
+  const [hotTable, setHotTable] = React.useState<HotTable | null>(null);
```

Reply:
side note: not mandatory, but we prefer to explicitly provide a default value for the useState. that makes these types of bugs a bit either to detect.

also, while we generally avoid null, in this case, the null is coming from react, not from us, and getting rid of null is more trouble than the benefits (would have to do `ref={(instance):void =>setHotTable(instance ?? undefined)}` to cast null to undefined. This is one of the few places where nulls are allowed in the codebase.
otherwise, we avoid them for consistency, to avoid bugs, and to have default arguments work correctly